### PR TITLE
Option Strategy: Jelly Roll & Ladder

### PR DESCRIPTION
**Pull request by LouisSzeto**
#### Description
Add new abstraction helper method and universe filter for option strategy Long/Short Jelly Roll, Bear/Bull Call/Put Ladder

Definition:
- Jelly Roll
  https://fincyclopedia.net/derivatives/l/long-jelly-roll
  https://www.sec.gov/files/rules/sro/phlx/2022/34-94293-ex5.pdf

- Ladder:
  https://en.wikipedia.org/wiki/Ladder_(option_combination)
  https://zerodha.com/varsity/chapter/bear-call-ladder/
  https://www.investopedia.com/terms/l/ladderoption.asp

margin requirement from IB:
- Long Jelly Roll
Screenshot 2024-06-14 163231

- Short Jelly Roll
Screenshot 2024-06-14 163428

- Bear Call Ladder
bear-call-ladder

- Bear Put Ladder
bear-put-ladder

- Bull Call Ladder
bull-call-ladder

- Bull Put Ladder
bull-put-ladder

#### Motivation and Context
no helper method for them

#### Requires Documentation Change
Add new section in Option Strategy

#### How Has This Been Tested?
- unit tests
- regression tests

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [X] I have added tests to cover my changes. 
- [X] All new and existing tests passed.
- [X] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Securities/Option/StrategyMatcher/OptionStrategyDefinitions.cs [hunk 1]
```diff
@@ -396,5 +396,92 @@ public static ImmutableList<OptionStrategyDefinition> AllDefinitions
...
+                OptionStrategyDefinition.PutLeg(-1, (legs, p) => p.Strike == legs[0].Strike,
+                                                    (legs, p) => p.Expiration == legs[0].Expiration),
+                OptionStrategyDefinition.PutLeg(+1, (legs, p) => p.Strike == legs[0].Strike,
+                                                    (legs, p) => p.Expiration == legs[1].Expiration)
+            );
+
+        /// <summary>
+        /// Bear Call Ladder strategy is short 1 call and long 2 calls, with ascending strike prices in order,
+        /// All options have the same expiry.
+        /// expiration.
```
Comment (Martin-Molinero) [inline Common/Securities/Option/StrategyMatcher/OptionStrategyDefinitions.cs:434]: Docs are off, please review all docs

File: Common/Securities/Option/OptionStrategies.cs [hunk 2]
```diff
@@ -912,6 +912,186 @@ public static OptionStrategy ShortBoxSpread(Symbol canonicalOption, decimal high
...
+        /// <param name="leg1Strike">The strike price of the short call</param>
+        /// <param name="leg2Strike">The strike price of the long call with lower strike price</param>
+        /// <param name="leg3Strike">The strike price of the long call with higher strike price</param>
+        /// <param name="expiration">Option expiration date</param>
+        /// <returns>Option strategy specification</returns>
+        public static OptionStrategy BearCallLadder(
+            Symbol canonicalOption,
+            decimal leg1Strike,
+            decimal leg2Strike,
+            decimal leg3Strike,
```
Comment (Martin-Molinero) [inline Common/Securities/Option/OptionStrategies.cs:966]: Believe `higherStrike/middleStrike/lowerStrike` would be better names, like we use in the other strategies, same comment for the other ones

File: Common/Securities/Option/StrategyMatcher/OptionStrategyDefinitions.cs [hunk 3]
```diff
@@ -396,5 +396,92 @@ public static ImmutableList<OptionStrategyDefinition> AllDefinitions
...
+            = OptionStrategyDefinition.Create("Bear Call Ladder",
+                OptionStrategyDefinition.CallLeg(-1),
+                OptionStrategyDefinition.CallLeg(+1, (legs, c) => c.Strike > legs[0].Strike,
+                                                     (legs, c) => c.Expiration == legs[0].Expiration),
+                OptionStrategyDefinition.CallLeg(+1, (legs, c) => c.Strike > legs[1].Strike,
+                                                     (legs, c) => c.Expiration == legs[0].Expiration)
+            );
+
+        /// <summary>
+        /// Bear Call Ladder strategy is long 1 put and short 2 puts, with descending strike prices in order,
```
Comment (Martin-Molinero) [inline Common/Securities/Option/StrategyMatcher/OptionStrategyDefinitions.cs:446]: > Bear Call Ladder

Off, please self review in detail

File: Common/Securities/Option/OptionStrategyPositionGroupBuyingPowerModel.cs [hunk 4]
```diff
@@ -246,6 +246,76 @@ public override MaintenanceMargin GetMaintenanceMargin(PositionGroupMaintenanceM
...
+
+                var strikeDifference = callSpread[1].Symbol.ID.StrikePrice - callSpread[0].Symbol.ID.StrikePrice;
+                var result = strikeDifference * multiplier;
+                var inAccountCurrency = parameters.Portfolio.CashBook.ConvertToAccountCurrency(result, shortCallSecurity.QuoteCurrency.Symbol);
+
+                return new MaintenanceMargin(result);
+            }
+            else if (_optionStrategy.Name == OptionStrategyDefinitions.BearPutLadder.Name)
+            {
+                // Bear Put Ladder = Bear Put Spread of 2 lower strike prices (margin: 0) + Short Naked Put with the highest strike price
```
Comment (Martin-Molinero) [inline Common/Securities/Option/OptionStrategyPositionGroupBuyingPowerModel.cs:278]: Seems these option strategies are actually a combination of others, we should reuse their margin requirement calculations and avoid having to duplicate the implementations, can extract implementation into some private method/abstraction if easier

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Cool! Leaving a few requests 👍

**Comment by Martin-Molinero** [comment]
Mind sharing where the option strategy definitions were sourced from, given they are not in the core IB list

**Comment by Martin-Molinero** [comment]
Please rebase from master, the regression test interface was changed & a unit test CI bug was shipped

**Comment by LouisSzeto** [comment]
> Mind sharing where the option strategy definitions were sourced from, given they are not in the core IB list

I've updated the description now 😄

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]

**Review by Martin-Molinero — APPROVED** [review]
Thank you 👍

**Pull request merged**